android-sdk/README.md

```diff
@@ -0,0 +1,224 @@
+
+Example registration of KeyCloakService
+```
+public class KeyCloakService implements ServiceModule {
+  static {
+    ServiceModuleRegistry.registerServiceModule("keycloak", KeyCloakService.class, "http");
+  }
+  /*snip*/
+}
+```
+
+ServiceModuleRegistry.java
+```
+
+/**
+ * Registry of service modules.  It is responsible for maintaining a mapping of service modules to
+ * their types as well as the dependencies each module expects.
+ */
+public final class ServiceModuleRegistry {
+    private static final Map<String, Class<? extends ServiceModule>> serviceTypeMap = new HashMap<>();
+    private static final Map<String, String> dependencyMap = new HashMap<>();
+    /**
+     * Register the looking for a service module to a class.
+     *
+     * @param type the type of module to use
+     * @param moduleClass the class that implements a service module type
+     * @param dependsOn the other serviceTypes this Service Depends on.
+     */
+    public static void registerServiceModule(String type, Class<? extends ServiceModule> moduleClass, String... dependsOn) {
+        serviceTypeMap.put(type, moduleClass);
+        for (String dependency : dependsOn) {
+            dependencyMap.put(type, dependency);
+        }
+    }
+
+}
+```
+** Discussion Point ** I'm not a fan of static methods during testing.  Perhaps there should be a method in the ServiceModule interface which returns a list of services that the service depends on.  MobileCore could then read that list by invoking that method as part of bootstrapping.
+
+#### Referencing a Service
+An instance of the core bassed during bootstrap can be used to reference other services.  MobileCore is responsible for ensuring that ServiceModule.bootstrap is not called until all declared dependencies are bootstrapped.
+
+SyncService.java
+```
+public class SyncService implements ServiceModule {
+
+  KeyCloakService keycloakService;
+  /*snip*/
+  public void bootstrap(MobileCore core, ServiceConfiguration config) {
+     this.keycloakService = core.getService("keycloak", KeyCloakService.class);
+     this.networkService = core.getService("network", NetworkService.class);
+     /*snip*/
+  }
+
+  public void performSync() {
+    if (keycloakService.isLoggedIn()) {
+      Reuqest request = networkService.newRequest();
+      request.addSecurityHeaders(keycloakService.getHeaders());//pseudocode method, don't expect this exactly.
+      request.setData(this.buildrequestBody());
+      Response response = networkService.execute(request);
+      response.onError(/*Error handling lambda*/);
+      response.onSuccess(/*Success handling lambda*/);
+    }
+  }
+
+}
+```
+
+#### Using Core Threads
+
+The Mobile Core SDK will provide Java Executors that wrap several common blocking tasks.  These executors will be available to any module or part of the application which may need to schedule diskIO work, netowrking work, or work to happen on the main thread.
+
+```
+AppExecutors.networkThread.execute(() -> {
```

Comment:
If we going to use OkHttpd.enqueue method then is that still needed? 
Requires investigation on how we can handle requests/library
Worried about application lifecycle.

Reply:
Yes it is still needed.